### PR TITLE
Enable nullability in MaskDescriptor

### PR DESCRIPTION
**Pull request by halgab**
## Proposed changes

- Enable nullability in `MaskDescriptor`

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9707)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDescriptorComparer.cs [hunk 1]
```diff
@@ -36,7 +34,7 @@ public int Compare(MaskDescriptor maskDescriptorA, MaskDescriptor maskDescriptor
             return 0;
         }
 
-        string textA, textB;
+        string? textA, textB;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDescriptorComparer.cs:37]: Please move `textB` to a separate line.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDescriptorTemplate.cs [hunk 2]
```diff
@@ -12,45 +10,39 @@ namespace System.Windows.Forms.Design;
...
 internal class MaskDescriptorTemplate : MaskDescriptor
 {
-    private readonly string _mask;
-    private readonly string _name;
-    private readonly string _sample;
-    private readonly Type _type;
-    private readonly CultureInfo _culture;
-
-    public MaskDescriptorTemplate(string mask, string name, string sample, Type validatingType, CultureInfo culture) : this(mask, name, sample, validatingType, culture, false)
+    public MaskDescriptorTemplate(string? mask, string name, string? sample, Type? validatingType, CultureInfo culture) : this(mask, name, sample, validatingType, culture, false)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDescriptorTemplate.cs:13]: Please add argument name before `false`

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs [hunk 3]
```diff
@@ -31,7 +31,7 @@ internal class MaskDesignerDialog : Form
     private ErrorProvider _errorProvider;
 
     private readonly List<MaskDescriptor> _maskDescriptors = new();
-    private MaskDescriptor _customMaskDescriptor;
+    private MaskDescriptorTemplate _customMaskDescriptor;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs:34]: Why did you change field type?

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs [hunk 4]
```diff
@@ -346,7 +346,7 @@ private bool ContainsMaskDescriptor(MaskDescriptor maskDescriptor)
         {
             Debug.Assert(descriptor is not null, "Null mask descriptor in the collection.");
 
-            if (maskDescriptor.Equals(descriptor) || maskDescriptor.Name.Trim() == descriptor.Name.Trim())
+            if (maskDescriptor.Equals(descriptor) || maskDescriptor.Name!.Trim() == descriptor.Name!.Trim())
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs:349]: Consider trimming and testing for null outside the loop, and return false if you got a null.
```
if (maskDescroptor is null || maskdescriptor.Name is null)
{
    return false;
}
```
and inside the loop - 
```
if (descriptor is null || descriptok.Name is null)
{
    return false;
}
```

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs:349]: Not sure I'm following here. All those can't be null, so why add the extra check? I'm moving the trimming outside the loop though

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs]: Sure. Should I explain it here or add a comment directly in the code?

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs]: We only add elements to the `_maskDescriptors` list

- at line 555, where we check `MaskDescriptor` for validity, which includes `Name` not being null
- at line 522, where we add `_customMaskDescriptor` whose `Name` we know is not null (because we created it)

**Comment by halgab** [comment]
It should be good now

**Review by lonitra — DISMISSED** [review]

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs (hunk 4 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs]: Thank you. Could you please add a comment in the code.

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/MaskDesignerDialog.cs]: I added an assert for `maskdescriptor.Name` and a comment for `descriptor.Name`

**Review by lonitra — APPROVED** [review]
